### PR TITLE
Criteo-audiences destination

### PR DESCRIPTION
**Pull request by pharika31**
_A summary of your pull request, including the what change you're making and why._

- Created new destination titled **'criteo-audiences'**
- Mandatory settings for the destination would be:
     - client_id
     - client_secret
     - advertiser_id
- Destination has the following actions:
     - **addUserToAudience** : This action would subscribe to type `'track'` and `'Audience Entered'` events from Personas payload. This action ultimately makes a PATCH request to Criteo API manage users endpoint to add emails to Criteo audience. 
     - **removeUserFromAudience** : This action would subscribe to type `'track'` and  `'Audience Exited'` events from Personas payload. This action ultimately makes a PATCH request to Criteo API manage users endpoint to remove emails from Criteo audience. 
- Action fields that we expect from the payload:
    - event
    - audience_key
    - email
- audience_key is being used to check if a corresponding Criteo audience exists with that name; otherwise we create a new audience_id on Criteo side and use that for subsequent path requests.
- Authentication scheme:  custom
   - We are using OAuth2.0 grant_type client_credentials flow. We make a request to get the access token for the PATCH request for the batch processing.

## Testing

_Include any additional information about the testing you have completed to
ensure your changes behave as expected. For a speedy review, please check
any of the tasks you completed below during your testing._

- [x] Added [unit tests](https://github.com/segmentio/action-destinations/blob/main/docs/testing.md#local-end-to-end-testing) for new functionality
- [x] Tested end-to-end using the [local server](https://github.com/segmentio/action-destinations/blob/main/docs/testing.md#local-end-to-end-testing)
- [ ] [Segmenters] Tested in the staging environment - unsure what this is about. Please let us know if we need to implement this.

## Comments/Questions on testing E2E:
- Testing with Payload consisting of 1000 events resulted in "payloadTooLarge" issue even though the size seemed <4MB limit.

### CODE REVIEW

File: packages/destination-actions/src/destinations/criteo-audiences/addUserToAudience/__tests__/index.test.ts [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+    const settings = VALID_SETTINGS;
+    nock('https://api.criteo.com').persist().post('/oauth2/token').reply(200, MOCK_TOKEN_RESPONSE)
+    nock('https://api.criteo.com').get(/^\/\d{4}-\d{2}\/audiences$/).query({ "advertiser-id": settings.advertiser_id }).reply(403)
+    await expect(
+      testDestination.testAction('addUserToAudience', {
+        event,
+        settings,
+        useDefaultMappings: true
+      })
+    ).rejects.toThrowError()
```
Comment (brennan) [inline packages/destination-actions/src/destinations/criteo-audiences/addUserToAudience/__tests__/index.test.ts:65]: When using `toThrowError()`, I think we should pass the error message as an argument so we validate that the desired error is what's being thrown. [Here's one example](https://github.com/segmentio/action-destinations/blob/8421315eacea3750822e5be81b2c859e542ad7b8/packages/destination-actions/src/destinations/salesforce/__tests__/sf-operations.test.ts#L75) of where we do this in our Salesforce integration.

File: packages/destination-actions/src/destinations/criteo-audiences/addUserToAudience/index.ts [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+import type { Settings } from '../generated-types'
+import type { Payload } from './generated-types'
+
+const getOperationFromPayload = async (
+  request: RequestFn,
+  advertiser_id: string,
+  payload: Payload[],
+  credentials: ClientCredentials
+): Promise<Operation> => {
+  const add_user_list: string[] = [];
```
Comment (brennan) [inline packages/destination-actions/src/destinations/criteo-audiences/addUserToAudience/index.ts:13]: Stylistic recommendation - semicolons at the end of lines aren't needed.

File: packages/destination-actions/src/destinations/criteo-audiences/addUserToAudience/index.ts [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+  settings: Settings,
+  payload: Payload[]
+): Promise<Response> => {
+  const credentials: ClientCredentials = {
+    client_id: settings.client_id,
+    client_secret: settings.client_secret
+  }
+  const operation: Operation = await getOperationFromPayload(request, settings.advertiser_id, payload, credentials);
+  return await patchAudience(request, operation, credentials)
+
```
Comment (brennan) [inline packages/destination-actions/src/destinations/criteo-audiences/addUserToAudience/index.ts:42]: Delete extra space.

File: packages/destination-actions/src/destinations/criteo-audiences/addUserToAudience/index.ts [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+import type { Payload } from './generated-types'
+
+const getOperationFromPayload = async (
+  request: RequestFn,
+  advertiser_id: string,
+  payload: Payload[],
+  credentials: ClientCredentials
+): Promise<Operation> => {
+  const add_user_list: string[] = [];
+  let audience_key = '';
```
Comment (brennan) [inline packages/destination-actions/src/destinations/criteo-audiences/addUserToAudience/index.ts:14]: Stylistic recommendation - Segment/Twilio typically uses `camelCase` for variable names in JavaScript, but using `snake_case isn't a blocker.

File: packages/destination-actions/src/destinations/criteo-audiences/addUserToAudience/index.ts [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+  for (const event of payload) {
+    if (!audience_key && event.audience_key)
```
Comment (brennan) [inline packages/destination-actions/src/destinations/criteo-audiences/addUserToAudience/index.ts:16]: I wonder if we should add a code comment here? This assumes that all events within the batch have the same `audience_key`. If I remember correctly, this is an assumption we agreed on. To start, customers should only connect a single audience to each instance of Criteo Audiences.

File: packages/destination-actions/src/destinations/criteo-audiences/addUserToAudience/index.ts (hunk 3 above)
Comment (pharika31) [reply packages/destination-actions/src/destinations/criteo-audiences/addUserToAudience/index.ts:42]: Edited.

File: packages/destination-actions/src/destinations/criteo-audiences/addUserToAudience/index.ts (hunk 4 above)
Comment (pharika31) [reply packages/destination-actions/src/destinations/criteo-audiences/addUserToAudience/index.ts:14]: As it's not a blocker, we shall keep snake_case for now.

File: packages/destination-actions/src/destinations/criteo-audiences/removeUserFromAudience/index.ts [hunk 6]
```diff
@@ -0,0 +1,87 @@
+import type { ActionDefinition } from '@segment/actions-core'
+import { getAudienceId, patchAudience } from '../criteo-audiences'
+import type { Operation, RequestFn, ClientCredentials } from '../criteo-audiences'
+import type { Settings } from '../generated-types'
+import type { Payload } from './generated-types'
+
+const getOperationFromPayload = async (
```
[the remark this answers is not in the document]
Comment (pharika31) [reply packages/destination-actions/src/destinations/criteo-audiences/removeUserFromAudience/index.ts:7]: `getOperationFromPayload` uses Payload[] interface parameter that we cannot import/use as an argument within criteo-audiences.ts.

File: packages/destination-actions/src/destinations/criteo-audiences/criteo-audiences.ts [hunk 7]
```diff
@@ -0,0 +1,192 @@
+import { IntegrationError } from '@segment/actions-core'
+import type { RequestClient } from '@segment/actions-core'
+
+const BASE_API_URL = 'https://api.criteo.com/2022-01'
+
+//export type RequestFn = (url: string, options?: RequestOptions) => Promise<Response>
```
Comment (brennan) [inline packages/destination-actions/src/destinations/criteo-audiences/criteo-audiences.ts:6]: Should we go ahead and remove this code?

File: packages/destination-actions/src/destinations/criteo-audiences/removeUserFromAudience/index.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (brennan) [reply packages/destination-actions/src/destinations/criteo-audiences/removeUserFromAudience/index.ts:9]: Ok, I think it's ok to leave separate. There's also the possibility that the `payload` parameter for the two actions may diverge in the future.

**Review by brennan — APPROVED** [review]

**Comment by brennan** [comment]
@pharika31 I'll plan to deploy this Monday, Pacific Time, so your team can begin end-to-end testing.

**Pull request merged**